Add HTTP-method-specific routing helpers

jankeromnes opened the pull request:
The following pull request makes it possible to do stuff like:

```
var camp = require('camp');
var app = camp.start();
app.get('/heeeeeeeey', function (req, res) {
  res.end('hooooooooo');
});
app.post('/secret', function (req, res) {
  console.log('not so secret:', req.data);
  res.end('done');
});
```

Additionally, it makes it possible to do `app.use(middleware)` (aliased to `app.handle(middleware)`). Both features make Camp's API a little bit more like Express's.

@espadrine please take a look.

jankeromnes commented on lib/camp.js:
```diff
@@ -351,6 +351,7 @@ function SecureCamp(opts) {
 inherits(SecureCamp, https.Server);
 
 Camp.prototype.handle = SecureCamp.prototype.handle =
+Camp.prototype.use = SecureCamp.prototype.use =
```
Middlewares are just functions handling `(augmentedRequest, augmentedResponse, next)`, so they should pretty much work the same for express and camp. Subtle differences are things like where parameters are located (e.g. for a path like `/users/:id`, express will have `req.params.id` vs camp's `req.data.id`), but then again some middlewares depend on other middlewares (e.g. using `req.cookies` in express assumes that you did something like `app.use(cookieParser())` before) so for more involved middlewares you always need to know what you're doing / what your `req` object has. TL;DR aliasing `handle` as `use` makes it clearer that it accepts middleware, and I guess it will simple work for most server middleware out there.

jankeromnes commented on lib/camp.js:
```diff
@@ -773,6 +774,38 @@ function pathLikeUnit(serverMethod) {
 var notFoundUnit = pathLikeUnit('notFound');
 
+// HTTP method unit (express-like, e.g. `camp.post('/', fn)`).
+function methodUnit (server) {
```
I agree it's better to support method-specific handlers in `pathUnit` directly, in order to preserve routing order.

Express only has single-method handlers (e.g. `app.get(path, (req, res) => {})`, `app.post(path, (req, res) => {})`, etc.) and one every-method handler that acts like a middleware (e.g. `app.all(path, (req, res, next) => {})`, so that you can let subsequent handlers catch your requests if you end up not handling them). I guess that's ok for most use cases. Do we really need N-method handlers that can set any number of methods?

jankeromnes commented on lib/camp.js:
```diff
@@ -773,6 +774,38 @@ function pathLikeUnit(serverMethod) {
+    if (!(path in paths)) {
+      paths[path] = {};
+      server._method(path, function (req, res, next) {
+        var cb = paths[path][req.method.toLowerCase()];
```
Indeed, `server._method` is implemented as a `pathLikeUnit`, so parametrized path matching is handled one layer above this. This `cb` holder is just there to match methods.

jankeromnes commented on lib/camp.js:
```diff
@@ -773,6 +774,38 @@ function pathLikeUnit(serverMethod) {
+    paths[path][method] = handler;
+  }
+
+  var methods = ['get', 'post', 'patch', 'put', 'delete'];
```
Indeed, express also implements a few more http://expressjs.com/en/guide/routing.html#route-methods. Maybe it makes sense to figure out the list of all methods we want to support.

jankeromnes commented:
Hi @espadrine, I addressed your comments and added a test. Please have another look.

Here is what I changed:

- Removed the `.use` alias for `.handle` that you didn't like (I'm doing it in [my own code](https://github.com/jankeromnes/janitor/blob/master/app.js#L40) anyway)
- Moved method routing into the path unit directly (callbacks can now be functions, or objects mapping methods to functions).
- It's quite easy to implement multi-method matching, by changing the `method` parameter of `addMethodCallback` to an array, but it would need further refactoring of the `pathLikeUnit` function (e.g. by adding yet another specific parameter like `serverMethodMethod = 'method'` or less barbaric name). I'm not sure how useful it is, but if you really want it, I'd like to open a follow-up bug for it. Not need to delay the basic helpers.
- The list of supported HTTP methods is now based on `http.METHODS`, i.e.:
  - acl
  - bind
  - checkout
  - connect
  - copy
  - delete
  - get
  - head
  - link
  - lock
  - m-search
  - merge
  - mkactivity
  - mkcalendar
  - mkcol
  - move
  - notify
  - options
  - patch
  - post
  - propfind
  - proppatch
  - purge
  - put
  - rebind
  - report
  - search
  - subscribe
  - trace
  - unbind
  - unlink
  - unlock
  - unsubscribe

jankeromnes commented:
Note: Travis fails because it uses `node v0.10.48`, whereas `http.METHODS` was introduced in `v0.11.8`: https://nodejs.org/api/http.html#http_http_methods

Maybe it's ok to upgrade `package.json` from `"engines": { "node": ">=0.10" }` to `"engines": { "node": ">=0.11.8" }`?

espadrine commented:
0.10 is ancient. It's OK to require 0.11.8.

espadrine commented on lib/camp.js:
```diff
@@ -768,18 +783,22 @@ function pathLikeUnit(serverMethod, statusCode) {
+            cb(req, res);
+          } else if (cb instanceof Object && cb[method] instanceof Function) {
+            cb[method](req, res);
           } else {
```
If it matches the path but not the method, it should `next()`.

espadrine commented on lib/camp.js:
```diff
@@ -768,18 +783,22 @@ function pathLikeUnit(serverMethod, statusCode) {
-          if (callbacks[cbindex] !== undefined) {
-            callbacks[cbindex](req, res);
+          var cb = callbacks[cbindex];
+          var method = req.method.toLowerCase();
```
Can we avoid that lowercase computation when this is not a method-specific filter?

espadrine commented on lib/camp.js:
```diff
@@ -739,18 +739,33 @@ function Path(matcher) {
+    server[serverMethod] = addCallback;
+
+    var addMethodCallback = function (method, path, callback) {
+      method = method.toLowerCase();
```
There is no need to lowercase it here.

espadrine commented on lib/camp.js:
```diff
@@ -739,18 +739,33 @@ function Path(matcher) {
+    };
+    httpMethods.forEach(function (method) {
+      method = method.toLowerCase();
+      server[method] = addMethodCallback.bind(null, method);
```
`method` must be lowercased for `server[method.toLowerCase()]`, but it is not needed elsewhere.

espadrine commented on lib/camp.js:
```diff
@@ -768,18 +783,22 @@ function pathLikeUnit(serverMethod, statusCode) {
+          var method = req.method.toLowerCase();
+          if (cb instanceof Function) {
+            cb(req, res);
+          } else if (cb instanceof Object && cb[method] instanceof Function) {
```
This polymorphism is not ideal. Instead, I would recommend to make `paths` an object containing all the filters, eg. `{path: new Path('/' + path), methods: […]}`.

Obviously, you'd have to change the name `paths` to something more appropriate. Maybe `filters`?

jankeromnes commented:
I rebased this pull request, and implemented all your suggestions (except a `server.method(['POST'], ...)` helper, because it would need yet another `pathLikeUnit` argument in order not to create a conflict on `server.method` for both `pathUnit` and `notFoundUnit`, could be done in a follow-up commit though).

@espadrine PTAL ("please take another look" 😄)

jankeromnes commented:
(Bumping minimum required node version to 0.11.8 for Travis.)

jankeromnes commented:
Note: I will refresh this pull request with optional parameters to `res.json()` so that implementers can choose whether they like to serve prettified JSON or not. This should be a good compromise to solve our disagreement in the other issue.

jankeromnes commented:
I rebased this pull request, and believe I found a good compromise for serving human-readable JSON: making it optional by passing `JSON.stringify()` parameters `replacer` and `space` to `res.json()` so that you can run `res.json(data, null, 2)`.

@espadrine, please have another look (all your previous suggestions should be implemented now, except the `server.method()` method, see earlier comments).

espadrine commented on lib/camp.js:
```diff
@@ -775,8 +795,9 @@ function pathLikeUnit(serverMethod, statusCode) {
-          if (callbacks[cbindex] !== undefined) {
-            callbacks[cbindex](req, res);
+          var cb = callbacks[cbindex] && callbacks[cbindex].callback;
+          if (cb) {
```
If possible, I would prefer to have booleans, instead of implicit conversion, in statement conditions.

espadrine commented on lib/camp.js:
```diff
@@ -745,25 +745,45 @@ function Path(matcher) {
-        matched = paths[i].match(req.path);
-        if (matched !== null) { cbindex = i; break; }
+        matched = callbacks[i].path.match(req.path);
+        if (!matched) {
```
If possible, I would prefer to have booleans, instead of implicit conversion, in statement conditions.

espadrine commented on lib/camp.js:
```diff
@@ -745,25 +745,45 @@ function Path(matcher) {
+          continue;
+        }
+        var methods = callbacks[i].methods;
+        if (methods && methods.indexOf(req.method) === -1) {
```
If possible, I would prefer to have booleans, instead of implicit conversion, in statement conditions.

espadrine commented on lib/camp.js:
```diff
@@ -745,25 +745,45 @@ function Path(matcher) {
-  statusCode = statusCode || 200;
+function pathLikeUnit(serverMethod, statusCode, httpMethods) {
+  httpMethods = httpMethods || [];
   return function unit(server) {
```
Not that important, but could you rename this function to `pathLikeUnit`? It eases debugging.

espadrine commented on lib/camp.js:
```diff
@@ -745,25 +745,45 @@ function Path(matcher) {
 }
 
 // Path unit.
```
Could you add a comment to warn that calling `pathLikeUnit()` with its third parameter more than once may reset `server[some method]`? Maybe describe the use of the `httpMethods` parameter.

jankeromnes commented:
@espadrine I fixed all your nits, please have another look when it's convenient for you. :)

espadrine commented:
Dandy!
